--!optimize 2
--!native
local packetIDs = require(script.Parent.packet_ids)
local readRefs = require(script.Parent.readRefs)

local ref = packetIDs.ref()
local free_thread: thread?

local function function_passer(fn, ...)
	local aquiredThread = free_thread
	free_thread = nil
	fn(...)
	free_thread = aquiredThread
end

local function yielder()
	while true do
		function_passer(coroutine.yield())
	end
end

local function run_listener(fn, ...)
	free_thread = task.spawn(free_thread or task.spawn(coroutine.create(yielder)), fn, ...)
end

local function read(incoming_buff: buffer, references: { [number]: unknown }?, player: Player?)
	local length = buffer.len(incoming_buff)
	local read_cursor = 0

	readRefs.set(references)

	while read_cursor < length do
		local packet = ref[buffer.readu8(incoming_buff, read_cursor)]
		read_cursor += 1

		local value, value_len = packet.reader(incoming_buff, read_cursor)

		read_cursor += value_len

		for _, listener in packet.getListeners() do
			run_listener(listener, value, player)
		end
	end
end

return read
